### PR TITLE
Find corners by equally distributed points

### PR DESCRIPTION
**Pull request by bramoo**
Remove duplicate peaks
Generate 4-sets of peaks
Evaluate sets with sum squared error from pi/2 radians

Works well on square pieces, which the pi/2 radian distribution assumes
Doesn't quite work on some particularly thin pieces, tested a bunch and found it missed the corners on 2, 11, and 204
Could probably improve by relaxing the pi/2 distribution to x, y, x, y where x + y + x + y = 2 pi

### CODE REVIEW

File: sovler.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        polar_piece_rho.append(rho)
+        polar_piece_phi.append(phi)
+
+    # Find peaks in doubled signal to get potential start/end peaks
+    peaks, _ = scipy.signal.find_peaks(polar_piece_rho + polar_piece_rho, prominence=1)
+
+    # Peaks are pulled from the doubled signal, get all peaks with index less
+    # than num_points and add peaks which wrap around but don't repeat past the
+    # first peak in the first copy of the signal
+    num_points = len(piece)
```
Comment (miclairy) [inline sovler.py]: I thought you could do this with
    deduplicated_peaks = {peak % len(piece_rho) for peak in peaks}

File: sovler.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    main_peaks = [p for p in peaks if p < num_points]
+    extra_peaks = [p % num_points for p in peaks if p >= num_points and p % num_points < min_peak]
+
+    deduped_peaks = extra_peaks + main_peaks
+
+    if len(deduped_peaks) < 4:
+        return ()
+
+    # Get all subsets of 4 peaks
+    combinations = list(itertools.combinations(deduped_peaks, 4))
```
Comment (miclairy) [inline sovler.py:76]: this is elegant

File: sovler.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+            score += ((math.pi / 2) - diff) ** 2
+        scores.append(score)
+
+    # Lowest score has phis distributed closest to pi/2 radians
+    best_combination_idx = scores.index(min(scores))
+    best_combination = combinations[best_combination_idx]
+
+    return best_combination
+
+def find_ins_outs(piece, corners):
```
Comment (miclairy) [inline sovler.py:97]: I thought about using distance to center based on corner locations

**Review by miclairy — APPROVED** [review]

File: sovler.py (hunk 1 above)
Comment (bramoo) [reply sovler.py]: 👍 yep much better

File: sovler.py (hunk 3 above)
Comment (bramoo) [reply sovler.py:97]: Yeah I'm trying some things here, I'm thinking if you look at distance from the edge defined by the corners, that will be a  mirror on the matching piece, whereas from the center won't be

**Pull request merged**